### PR TITLE
Adding "mutable log paths" feature 

### PR DESCRIPTION
**Pull request by galmasi**
As discussed in Issue #657  -- initial commit for implementing configurable logging paths

### CODE REVIEW

**Comment by galmasi** [comment]
As mentioned above, please do not merge yet -- this is a trial balloon to see how many testing features I run afoul in one try. Also, the implementation is incomplete: the feature discussed in Issue #657 , info logging nonstandard log paths, is not yet implemented.

**Comment by galmasi** [comment]
@ansasaki @kkaarreell I do not believe I can improve the rust agent code in this PR anymore. I have been laboring under a misunderstanding of how rust agent testing is performed. I have created this one-line [PR] (https://github.com/RedHat-SP-Security/keylime-tests/pull/498) in `keylime-tests` to help pass this PR.

Note that the `testing` feature in the keylime agent is not actually used for measured boot in the test framework; instead, the test framework directly `sed`s the source code. This raises the issue that either 

(a) the testing harness should be updated to compile the rust agent correctly by using cargo with the `--feature testing` parameters; -- or -- 
(b) let the SED command stay in the test code and remove the unused `testing` feature of the keylime agent.

**Comment by ansasaki** [comment]
> @ansasaki @kkaarreell I do not believe I can improve the rust agent code in this PR anymore. I have been laboring under a misunderstanding of how rust agent testing is performed. I have created this one-line [PR] ([RedHat-SP-Security/keylime-tests#498](https://github.com/RedHat-SP-Security/keylime-tests/pull/498)) in `keylime-tests` to help pass this PR.
> 
> Note that the `testing` feature in the keylime agent is not actually used for measured boot in the test framework; instead, the test framework directly `sed`s the source code. This raises the issue that either
> 
> (a) the testing harness should be updated to compile the rust agent correctly by using cargo with the `--feature testing` parameters; -- or -- (b) let the SED command stay in the test code and remove the unused `testing` feature of the keylime agent.

This looks like a leftover. It should use the approach `a` and avoid changing the code. I'll try to get rid of that `sed`.

**Comment by galmasi** [comment]
I closed https://github.com/RedHat-SP-Security/keylime-tests/pull/498 in the expectation that https://github.com/RedHat-SP-Security/keylime-tests/pull/499 will solve this problem. Will force a re-test as soon as 499 is merged.

File: keylime-agent/src/config.rs
Comment (ansasaki) [inline keylime-agent/src/config.rs:739]: I think using `config_get_file_path()` is not correct here. Note that it will use the provided default path relative from the `keylime_dir` (by default `/var/lib/keylime`).

Maybe a way to fix this would be to change the `DEFAULT_AGENT_DATA_PATH`, `DEFAULT_SERVER_KEY`, `DEFAULT_SERVER_CERT`, etc. to be relative paths (i.e. not starting with `/`) and make `config_get_file_path`  to prepend the `keylime_dir` only when the default value is a relative path.

File: keylime-agent/src/main.rs
Comment (ansasaki) [inline keylime-agent/src/main.rs:187]: This should be dropped, and the tests modified to use the new added way to set the measurements logs.

Note that the tests only passed because this is still in place, as the CI is setting this env var.  Otherwise it would had failed as the resulting path after `config_get_file_path()` would not be the expected one.

**Review by ansasaki — CHANGES_REQUESTED** [review]

File: GNUmakefile
Comment (galmasi) [reply GNUmakefile:28]: Double quote means that multiple clean targets can be attached. It's good `Makefile` practice, esp. for `clean` because there might be multiple aspects to clean up.

+1 I will add the `.PHONY` notation.

File: keylime-agent/src/main.rs
Comment (galmasi) [reply keylime-agent/src/main.rs:187]: I'd be very happy to drop this, but not while the CI is using it -- correct?
Unless you want to hold this PR until CI is fixed, we need to leave it in place.

Comment (ansasaki) [reply keylime-agent/src/main.rs:187]: This can be fixed in a separate PR.

File: keylime-agent/src/config.rs
Comment (ansasaki) [reply keylime-agent/src/config.rs:739]: Alternatively, I think it is acceptable that instead of `keylime_dir`, the `/` directory is used in this case.

Also, please add the new options to  `keylime-agent.conf` with a description and default value set. The configuration minor version number should be bumped.

**Comment by galmasi** [comment]
I'll take another look at all the changes I have made and try to reconcile. I have rebased to the current master, but there are a number of thinking errors that I need to address: chief among them, `config_file_get_path()`. I need to build a test case that shows it used wrong, then fix it.

File: keylime-agent/src/config.rs
Comment (galmasi) [reply keylime-agent/src/config.rs:758]: "/" is fine. I was thinking of "/sys/kernel/security", just in case anyone would ever want to specify the IMA or MB log file as a relative path. But that's not worth parametrizing, so probably "/" is safer and less involved. OK, let's do that.

**Review by ansasaki — APPROVED** [review]

File: keylime-agent/src/config.rs
Comment (aplanas) [inline keylime-agent/src/config.rs:535]: So many `_` seems a bit ugly. I wonder if this can help: https://github.com/rust-lang/rfcs/issues/3092

**Review by aplanas — APPROVED** [review]

**Comment by ansasaki** [comment]
Hi @galmasi, I took the liberty to make the formatting changes and rebase. I'd like to ask if I can also add the configuration change required (in the case, include the new configuration options and bump the version)

File: keylime-agent/src/config.rs
Comment (ansasaki) [reply keylime-agent/src/config.rs:535]: Thanks for the pointer, but unfortunately we cannot use this (yet). It is still available only on nightly:
https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.try_insert

**Comment by ansasaki** [comment]
I'll squash some commits and rebase. Note that I added a commit to add the new configuration options.

**Pull request merged**
